src/lib/transition_frontier_controller_tests/test_catchup_scheduler.ml

```diff
@@ -0,0 +1,248 @@
+        Strict_pipe.create Synchronous
+      in
+      Thread_safe.block_on_async_exn (fun () ->
+          let open Deferred.Let_syntax in
+          let%bind frontier = setup_random_frontier () in
+          let scheduler =
+            Catchup_scheduler.create ~logger ~frontier ~time_controller
+              ~catchup_job_writer ~catchup_breadcrumbs_writer
+          in
+          let randomly_chosen_breadcrumb =
+            Transition_frontier.all_breadcrumbs frontier
+            |> List.permute |> List.hd_exn
+          in
+          let%bind upcoming_breadcrumbs =
+            Deferred.all
+            @@ Quickcheck.random_value
+                 (gen_linear_breadcrumbs ~logger ~size:2
+                    ~accounts_with_secret_keys randomly_chosen_breadcrumb)
+          in
+          let missing_hash =
+            List.hd_exn upcoming_breadcrumbs
+            |> Transition_frontier.Breadcrumb.transition_with_hash
+            |> With_hash.hash
+          in
+          let dangling_breadcrumb = List.nth_exn upcoming_breadcrumbs 1 in
+          let dangling_transition =
+            Transition_frontier.Breadcrumb.transition_with_hash
+              dangling_breadcrumb
+          in
+          Catchup_scheduler.watch scheduler ~timeout_duration
+            ~transition:dangling_transition ;
+          let result_ivar = Ivar.create () in
+          Strict_pipe.Reader.iter_without_pushback catchup_job_reader
+            ~f:(fun catchup_hash -> Ivar.fill result_ivar catchup_hash )
+          |> don't_wait_for ;
+          let%map catchup_hash = Ivar.read result_ivar in
+          assert (Coda_base.State_hash.equal missing_hash catchup_hash) ;
+          Strict_pipe.Writer.close catchup_breadcrumbs_writer ;
+          Strict_pipe.Writer.close catchup_job_writer )
+
+    let%test_unit "if a linear sequence of transitions in reverse order, \
+                   catchup scheduler should not create duplicate jobs" =
+      let _catchup_job_reader, catchup_job_writer =
+        Strict_pipe.create Synchronous
+      in
+      let catchup_breadcrumbs_reader, catchup_breadcrumbs_writer =
+        Strict_pipe.create Synchronous
+      in
+      Thread_safe.block_on_async_exn (fun () ->
+          let open Deferred.Let_syntax in
+          let%bind frontier = setup_random_frontier () in
+          let scheduler =
+            Catchup_scheduler.create ~logger ~frontier ~time_controller
+              ~catchup_job_writer ~catchup_breadcrumbs_writer
+          in
+          let randomly_chosen_breadcrumb =
+            Transition_frontier.all_breadcrumbs frontier
+            |> List.permute |> List.hd_exn
+          in
+          let%bind upcoming_breadcrumbs =
+            Deferred.all
+            @@ Quickcheck.random_value
+                 (gen_linear_breadcrumbs ~logger ~size:4
+                    ~accounts_with_secret_keys randomly_chosen_breadcrumb)
+          in
+          let missing_breadcrumb = List.nth_exn upcoming_breadcrumbs 0 in
```

Comment:
This following statements is quite verbose. You can use `List.fold_right` to make this more simple.